internal/pkg/build/args/reader.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-2023, Sylabs Inc. All rights reserved.
+// Copyright (c) Contributors to the Apptainer project, established as
+//   Apptainer a Series of LF Projects LLC.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package args
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+
+	"github.com/samber/lo"
+	"github.com/sylabs/singularity/pkg/sylog"
+)
+
+type Reader struct {
+	src            *bufio.Reader
+	buildArgsMap   map[string]string
+	defaultArgsMap map[string]string
+	consumedArgs   map[string]bool
+	buf            bytes.Buffer
+	bufReader      io.Reader
+	bufWriter      io.Writer
+}
+
+const bufSize = 1024
+
+var (
+	buildArgsRegexp              = regexp.MustCompile(`{{\s*(\w+)\s*}}`)
+	buildArgsIncompleteSrcRegexp = regexp.MustCompile(`{(({\s*((\w+\s*(}?))?))?)$`)
+)
+
+func NewReader(src io.Reader, buildArgsMap map[string]string, defaultArgsMap map[string]string) *Reader {
+	r := &Reader{
+		src:            bufio.NewReader(src),
+		buildArgsMap:   buildArgsMap,
+		defaultArgsMap: defaultArgsMap,
+	}
+
+	r.consumedArgs = make(map[string]bool)
+	r.bufReader = io.Reader(&r.buf)
+	r.bufWriter = io.Writer(&r.buf)
+
+	return r
+}
+
+func (r *Reader) Read(p []byte) (n int, err error) {
+	var srcBytes []byte
+	var matches [][]int
+	atRealEOF := false
+	for !atRealEOF {
+		srcBufSize := r.src.Buffered()
+		sylog.Debugf("Underlying bufio.Reader has %d buffered bytes", srcBufSize)
```

Comment:
This level of debug message is too chatty for the end-user.

Ideally this Read function is proven working by unit tests. It doesn't interact with the external system, so we shouldn't need this level of info shown to the user.